util/offline_db/get_resources_from_db.py

```diff
@@ -0,0 +1,262 @@
+    }
+    data = json.dumps(data).encode("utf-8")
+    headers = {
+        "Authorization": f"Bearer {token}",
+        "Content-Type": "application/json",
+    }
+
+    req = request.Request(url, data=data, headers=headers)
+
+    try:
+        response = request.urlopen(req)
+    except Exception as e:
+        raise Exception(
+            "Failed to obtain resources via url request. "
+            f"Exception thrown:\n{e}"
+        )
+
+    json_response = json.loads(response.read().decode("utf-8"))
+    assert "documents" in json_response, (
+        "'documents' not a key in "
+        "json dictionary. JSON: \n"
+        f"{json_response}"
+    )
+    resources = json_response["documents"]
+
+    return resources
+
+
+def save_resources_to_file(resources: List[Dict], output: TextIO):
+    """
+    This function saves all the JSON objects for resources to a file.
+    :param resources: List of JSON objects for resources
+    :param output: Output directory absolute path
+    """
+
+    path = output.joinpath("resources.json")
+
+    with open(path, "w") as f:
+        json.dump(resources, f, indent=4)
+
+
+def get_resource_local_filepath(
+    resource_url: str, base_output_path: TextIO
+) -> Path:
+    """
+    This function returns the local filepath for a resource
+    :param resource_url: URL field of the resource
+    :param base_output_path: Base output directory absolute path
+    :return: Local filepath for the resource
+    """
+    # This line get the file path that is in the  google
+    # cloud storage bucket after the ``dist`` folder.
+    # The url: https://storage.googleapis.com/dist.gem5.org/dist/develop/checkpoints/...
+    # will be converted to  develop/checkpoints/...
+    filename = Path(
+        # This line combines the parts of the path
+        *Path(
+            # This line gets the path after https://
+            parse.urlsplit(resource_url).path
+        )
+        # This line splits the path into parts
+        # and gets the parts after the dist folder
+        .parts[3:]
+    )
+
+    filepath = Path(base_output_path).joinpath(filename)
+
+    filepath.parent.mkdir(parents=True, exist_ok=True)
+
+    return filepath
+
+
+def download_resources(resources: List[Dict], output: TextIO):
+    """
+    This function downloads the resources which have a url field and
+    updates the url field to point to the local download of the resource.
+    :param output: Output directory absolute path
+    """
+
+    path = output.joinpath("offline_resources")
+    if not path.exists():
+        path.mkdir()
+
+    for resource in resources:
+        if "url" in resource.keys():
+            url = resource["url"]
+            filepath = get_resource_local_filepath(url, path)
+
+            if (
+                not filepath.exists()
+                or "url-md5sum" not in resource.keys()
+                or hashlib.md5(filepath.read_bytes()).hexdigest()
+                != resource["url-md5sum"]
+            ):
```

Comment:
I think the logic here is slightly incorrect. In this logic the file will be redownloaded everytime if `url-md5sum` does not exist e. I think the following is what you want:

```suggestion
            if (
                not filepath.exists() # If the file does not exist
                or ("url-md5sum" in resource.keys() # If it exist but the md5sum is wrong.
                and hashlib.md5(filepath.read_bytes()).hexdigest()
                != resource["url-md5sum"])
            ):
```